src/gui/gt_mdiitem.cpp

```diff
@@ -18,12 +18,48 @@
 #include "gt_projectchangedevent.h"
 #include "gt_objectchangedevent.h"
 #include "gt_statehandler.h"
+#include "gt_qmlaction.h"
+
+
+struct GtMdiItem::Impl
+{
+    Impl() :
+        m_subWin(nullptr),
+        m_queueEvents(false)
+    {}
+
+    ~Impl()
+    {
+        qDeleteAll(m_eventQueue);
+    }
+
+    QWidget* m_subWin;
+    QList<GtQueuedMdiEvent*> m_eventQueue;
+    bool m_queueEvents;
+
+    /// The list of registered toolbar actions, non-owning
+    std::vector<GtQmlAction*> m_toolbarActions;
+
+    template <class T>
+    T* takeEvent()
+    {
+        foreach (GtQueuedMdiEvent* e, m_eventQueue)
+        {
+            if (T* ce = qobject_cast<T*>(e))
+            {
+                m_eventQueue.removeOne(e);
+                return ce;
+            }
+        }
+
+        return NULL;
```

Comment:
I think this should work and avoids the copy of foreach. But I'm not 100% certain.
 ```suggestion
        auto iter = std::find_if(m_eventQueue.begin(), 
                                 m_eventQueue.end(), 
                                 [](GtQueuedMdiEvent* e){
            return qobject_cast<T*>(e);
        });
        if (iter != m_eventQueue.end()) return nullptr;
        
        T* ce = static_cast<T*>(*iter);
        m_eventQueue.erase(iter);
        return ce;
```

Reply:
This was moved code into the pimpl.

Reply:
I see, but maybe you could touch it up a little while you are at it? Or should we leave it as is?

Reply:
I have thought about it. Your solution does not work and you need to cast twice, which makes the code more complicated as before. We can still avoid the copy by using a normal for loop instead of a foreach